akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorSpec.scala

```diff
@@ -697,7 +716,88 @@ class EventSourcedBehaviorSpec extends ScalaTestWithActorTestKit(EventSourcedBeh
+      persistentActor ! GetValue(replyProbe.ref)
+      replyProbe.expectMessage(State(3, Vector(0, 1, 2)))
+      snapshotProbe.expectMessage(Try(Done))
+      retentionProbe.expectMessageType[Success[DeleteSnapshotsCompleted]].value match {
+        case DeleteSnapshotsCompleted(Criteria(SnapshotSelectionCriteria(maxSequenceNr, _, minSequenceNr, _))) =>
+          maxSequenceNr shouldEqual 2
+          minSequenceNr shouldEqual 0
+        case signal => unexpected(signal)
+      }
+
+      persistentActor ! IncrementWithPersistAll(3)
+      snapshotProbe.expectMessage(Try(Done))
+      retentionProbe.expectMessageType[Success[DeleteSnapshotsCompleted]].value match {
+        case DeleteSnapshotsCompleted(Criteria(SnapshotSelectionCriteria(maxSequenceNr, _, minSequenceNr, _))) =>
+          maxSequenceNr shouldEqual 5
+          minSequenceNr shouldEqual 1
+        case signal => unexpected(signal)
+      }
+
+      persistentActor ! GetValue(replyProbe.ref)
+      replyProbe.expectMessage(State(6,Vector(0, 1, 2, 3, 4, 5)))
+    }
+
+    "optionally delete both old messages and snapshots" in {
+      val pid = nextPid
+      val snapshotProbe = TestProbe[Try[Done]]()
+      val retentionProbe = TestProbe[Try[RetentionSignal]]()
+      val replyProbe = TestProbe[State]()
+
+      val persistentActor = spawn(Behaviors.setup[Command](ctx ⇒
+        counterWithSnapshotAndRetentionProbe(ctx, pid, snapshotProbe.ref, retentionProbe.ref)
+          .snapshotEvery(2)
+          .withRetention(
+            snapshotEveryNEvents = 2,
+            keepNSnapshots = 2,
+            deleteMessagesOnSnapshot = Some(true))))
+
+      persistentActor ! IncrementWithPersistAll(10)
+      persistentActor ! GetValue(replyProbe.ref)
+      val initialState = replyProbe.expectMessageType[State]
+      initialState.value shouldEqual 10
+      initialState.history shouldEqual (0 until initialState.value).toVector
+      snapshotProbe.expectMessage(Try(Done))
+
+      val firstDeleteMessages = retentionProbe.expectMessageType[Success[DeleteMessagesCompleted]].value
+      firstDeleteMessages.toSequenceNr shouldEqual 6 // 10 - 2 * 2
+
+      retentionProbe.expectMessageType[Success[DeleteSnapshotsCompleted]].value match {
+        case DeleteSnapshotsCompleted(Criteria(SnapshotSelectionCriteria(maxSequenceNr, _, minSequenceNr, _))) =>
+          maxSequenceNr shouldEqual 5 // 10 / 2
+          minSequenceNr shouldEqual 1
+        case signal => fail(s"Unexpected signal [$signal].")
+      }
+
+      persistentActor ! IncrementWithPersistAll(10)
+      snapshotProbe.expectMessage(Try(Done))
+      val secondDeleteMessages = retentionProbe.expectMessageType[Success[DeleteMessagesCompleted]].value
+      secondDeleteMessages.toSequenceNr shouldEqual 16 // 20 - 2 * 2
+
+      persistentActor ! GetValue(replyProbe.ref)
+      val state = replyProbe.expectMessageType[State]
+      state.value shouldEqual 20
+      state.history shouldEqual (0 until state.value).toVector
```

Comment:
Maybe good to add a test where it is actually slow rather than completing immediately also?

Reply:
There's two tests I'd like to add but I was hoping to do so not in this commit if that's acceptable. Your suggestion and a larger number of messages case.

Reply:
Creating separate ticket - a one-off, would not usually do.